File: src/main/scala/marksync/FileInfo.scala
package marksync

import java.io.File
import java.nio.file.Files
import java.security.MessageDigest

import org.apache.commons.codec.binary.Hex

case class FileInfo(filename: String, digest: String, url: Option[String]) {
  def isIdenticalTo(file: File): Boolean = digest == FileInfo.sha1(file)
}

object FileInfo {
  def apply(filename: String, file: File, url: Option[String]): FileInfo =
    FileInfo(filename, sha1(file), url)

  def sha1(file: File): String = {
    val digest = MessageDigest.getInstance("SHA-1")
    digest.update(Files.readAllBytes(file.toPath))
    Hex.encodeHexString(digest.digest)
  }
}

File: src/main/kotlin/marksync/uploader/FileInfo.kt
package marksync.uploader

import org.apache.commons.codec.binary.Hex
import java.io.File
import java.nio.file.Files
import java.security.MessageDigest

data class FileInfo(
    val filename: String,
    val digest: String?,
    val url: String?
) {
    constructor(filename: String, file: File, url: String?) :
            this(filename, sha1(file), url)

    /**
     * Check if this file is not modified.
     */
    fun isIdenticalTo(file: File): Boolean = digest == sha1(file)

    companion object {
        /**
         * Calculate file digest.
         */
        fun sha1(file: File): String {
            val digest = MessageDigest.getInstance("SHA-1")
            digest.update(Files.readAllBytes(file.toPath()))
            return Hex.encodeHexString(digest.digest())
        }
    }
}
